src/core/include/mp-units/numbers.h

```diff
@@ -0,0 +1,299 @@
+
+template<typename T, typename U, typename V>
+concept multiplication_with =  // clang-format off
+  number<T> &&
+  number<U> &&
+  number<V> &&
+  requires( const T & c,  const U & u) {
+    { (c * u) } -> common_number_with<V>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_multiplication_with =  // clang-format off
+  multiplication_with<T,U,T> &&
+  requires(T& l,  const U & u) {
+    { l *= u } -> std::same_as<T&>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept division_with =  // clang-format off
+  multiplication_with<T,U,T> &&
+  multiplication_with<U,T,T> &&
+  requires( const T & c,  const U & u) {
+    { c / u } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_division_with =  // clang-format off
+  division_with<T,U> &&
+  compound_multiplication_with<T,U> &&
+  requires(T& l,  const U & u) {
+    { l /= u } -> std::same_as<T&>;
+  };  // clang-format on
+
+}  // namespace detail
+
+template<typename T, typename U>
+concept modulo_with =  // clang-format off
+  number<T> &&
+  number<U> &&
+  requires( const T & c,  const U & u) {
+    { (c % u) } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_modulo_with =  // clang-format off
+  modulo_with<T,U> &&
+  requires(T& l,  const U & u) {
+    { l %= u } -> std::same_as<T&>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept modulus_for = modulo_with<U, T>;
+template<typename T, typename U>
+concept compound_modulus_for = compound_modulo_with<U, T>;
+
+template<typename T>
+concept negative =  // clang-format off
+  detail::compound_addition_with<T,T> &&
+  requires( const T & c) {
+    number_zero_v<T>;
+    { -c } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T>
+concept set_with_inverse =  // clang-format off
+  detail::compound_multiplication_with<T,T> &&
+  requires( const T & c) {
+    { number_one_v<T> / c } -> std::common_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept point_space_for =
+  detail::subtraction_with<T, U> && negative<U> && common_number_with<number_difference_t<T>, U>;
+template<typename T, typename U>
+concept compound_point_space_for = point_space_for<T, U> && detail::compound_subtraction_with<T, U>;
+template<typename T>
+concept point_space = compound_point_space_for<T, number_difference_t<T>>;
+
+template<typename T, typename U>
+concept vector_space_for = point_space_for<U, T>;
+template<typename T, typename U>
+concept compound_vector_space_for = compound_point_space_for<U, T>;
+
+namespace detail {
+
+template<typename T>
+concept weak_scalar = common_number_with<T, number_difference_t<T>> && point_space<T> && negative<T>;
+
+template<typename T, typename U>
+concept scales_with =
+  common_number_with<U, vector_scalar_t<T>> && weak_scalar<U> && multiplication_with<T, U, T> && set_with_inverse<U>;
```

Comment:
Are you saying that dividing is a compile-time error if the divisor has `0` in its range?

Reply:
For `safe::var`, that's exactly the case.

Reply:
```c++
using namespace safe::literals;

// declare variable `x` to be between -10 and 10 inclusive, guaranteed at compile time:
safe::ival_s32<-10, 10> x = 10_s32;
safe::ival_s32<1, 10> y = 10_s32;

// it's possible to have a number type that excludes '0', but still has positive and negative values
safe::var<int32_t, safe::ival<-10, -1> || safe::ival<1, 10>> z = 1_s32;

// these are fine:
auto a = x / y;
auto b = x / z;

// this is not fine and will cause a static_assert, because x _might_ be zero:
auto c = y / x;

```

https://github.com/intel/safe-arithmetic

Reply:
I think the problem is that _`scales-with`_`<T, U>` shouldn't require `set_with_inverse<U>`.
I think I should have moved it to `field_for` and `compound_field_for`.